Feature global pointers

roigcarlo opened the pull request:
This PR add global pointers. They will be used in a near future

roigcarlo committed:
Cleanup

roigcarlo committed:
Removed test to check the life cycle of the global+boost pointer interaction since we decided not to explicitly maintain this interaction

roigcarlo committed:
Adding test to check invalid acces while running with no mpi

roigcarlo committed:
return rank not size

roigcarlo committed:
Using TData instead of TDataPtr

roigcarlo committed:
Disable rank test and updated existing to use data type instead of data pointer

pooyan-dadvand committed:
Changing the deleted constructor to be pubiic

pooyan-dadvand committed:
Adding the constant operator and other minor changes

roigcarlo committed:
Renamed Kratos_global_pointer to global_pointer

roigcarlo committed:
Remove unnecessary include

roigcarlo committed:
Reming rank from non-mpi code

roigcarlo committed:
Typo

roigcarlo committed:
Removing unused tests

roigcarlo committed:
added const tests

roigcarlo committed:
Updated doc

roigcarlo committed:
Added some deleted constructors to prevent problems with boost and std smart pointers

roigcarlo committed:
Added additional delete constructor for base datatype

roigcarlo committed:
Added to/from raw data methods

roigcarlo committed:
Added mpi gather tests

roigcarlo committed:
Changed from/to raw to load/save and simplified

RiccardoRossi commented on kratos/includes/global_pointer.h:
```diff
@@ -0,0 +1,186 @@
+   */
+  GlobalPointer(TDataType * DataPointer)
+    : mDataPointer(DataPointer)
+#ifdef KRATOS_USING_MPI
```
this makes impossible to use in the same executable MPI and OpenMP alone (without MPI).

i would love to keep that possibility... any way to prserve that?
to explain myself more precisely, the scenario i see is the following:

i am on a cluster, and i compile with mpi=0n. Now i want to run OpenMP alone. I will still need mpirun...while this is definitely not what i would want. An excellent example of this is to run the tests

RiccardoRossi commented on kratos/includes/global_pointer.h:
```diff
@@ -0,0 +1,186 @@
+   * @return rank of the data owner
+   */
+  int GetRank() const {
+#ifdef KRATOS_USING_MPI
```
same comment as above...

RiccardoRossi replied:
@pooyan-dadvand @jcotela 

i am ok with having rank in the kernel, however i would call it something like "unique_processor_id" or smthg of the sort.

my point is that one may have different modelparts employing different communicator worlds

pooyan-dadvand replied:
@RiccardoRossi the unique process id would mix the MPI and openMP?

philbucher commented:
@roigcarlo are you planning on implementing a parallel data exchange for this?

roigcarlo commented:
@philbucher Yes and no. The final goal is actually a search algorithm in which almost all the data is kept in the owner process. Think of it a pseudo-mapreduceish kind of scheme.

At some point you will have a container of global pointers in which you have a function to apply and a "worker" that will distribute the work based on where the pointers are located.

Veeery schematically:

```C++
void DoOperation(Data) {
   ....
}

Apply(vector<GlobalPointer> Data, Function=DoOperation) {
    DataProc = GroupByProcID(Data) 

    for(every proc) {
        if local_proc
           DoOperation(DataProc[thisProc])
        else
           SendData(DataProc[thisProc])
    }
}
```

So you will basically have a pointer in your process which points to data in other process, but you will be able to operate normally via the worker and perform simple operations with it ( count, move, etc...)

pooyan-dadvand commented on kratos/includes/global_pointer.h:
```diff
@@ -0,0 +1,186 @@
+   * Constructor by std::shared_ptr
+   * @param DataPointer Std Shared Pointer to the Data.
+   */
+  GlobalPointer(std::shared_ptr<TDataType> DataPointer) = delete;
```
I believe that we should do the same as boost::share_ptr

pooyan-dadvand commented on kratos/includes/global_pointer.h:
```diff
@@ -0,0 +1,186 @@
+   * Constructor by boost::weak_ptr
+   * @param DataPointer Boost Weak Pointer to the Data.
+   */
+  GlobalPointer(boost::weak_ptr<TDataType> DataPointer) = delete;
```
Is there any specific reason to not create it via a weak pointer?

roigcarlo replied:
I suppose no. anyway, its odd since this IMHO can be seen from two different sides:

1 - We use the current implementation of the global_ptr:
It means that the lock needs to be done when acquiring the data, so once the user creates the global_ptr, the old weak_ptr is useless, as one can easily bypass the lock by accessing the data directly from the global_ptr.

2 - We store the weak_ptr inside the global_ptr:
Then the lock can be made when "using" the pointer, but it needs a completely separate implementation in order to ensure that honors the lock mechanism.

I uploaded following 1, but I would not recommend using this...

pooyan-dadvand commented on kratos/includes/global_pointer.h:
```diff
@@ -0,0 +1,186 @@
+
+} // namespace Kratos
+
+#endif // KRATOS_GLOBAL_POINTER_H_INCLUDED
```
Would you please add the copy constructor, move constructor and copy and move = operator?

pooyan-dadvand commented on kratos/tests/global_pointers/test_global_pointer.cpp:
```diff
@@ -0,0 +1,174 @@
+{
+  int sampleVar = 1337;
+
+	auto fromRaw = GlobalPointer<int>(&sampleVar);
```
local variables should be like form_raw

pooyan-dadvand commented on kratos/tests/global_pointers/test_global_pointer.cpp:
```diff
@@ -0,0 +1,174 @@
+/// Parallel tests
+
+#ifdef KRATOS_USING_MPI
+KRATOS_TEST_CASE_IN_SUITE(GlobalPointerGatherRaw, KratosCoreFastSuit)
```
I think we should add a macro for MPI test cases which deos the enable disable. But it would be subject of another branch. It is ok as you did now.

pooyan-dadvand commented on kratos/tests/global_pointers/test_global_pointer.cpp:
```diff
@@ -0,0 +1,174 @@
+
+  std::size_t gpSize = sizeof(GlobalPointer<int>);
+
+  char * rawGatherSend = (char *)malloc(gpSize);
```
It would be better to use new and delete

pooyan-dadvand reviewed (commented):
It is ok and aligend with our design! good job!

Nevertheless there are some minor improvements

philbucher commented:
@roigcarlo this is what I meant, will this worker do data exchange btw ranks in the background? In your example the `Apply` function would be the worker that also exchanges the data?
Sorry closed by accident

roigcarlo commented:
@philbucher  Yep exactly

philbucher commented:
@roigcarlo and the `Apply` function would be provided or implemented ourselves?

roigcarlo commented:
Apply function will be provided by us. but I don't want to advance things since we still have to discuss with @pooyan-dadvand implementation details

philbucher commented:
ok thx for the explanation

roigcarlo committed:
Added constructor for std::shared_ptr

roigcarlo committed:
Naming convetions

roigcarlo committed:
missing operators

roigcarlo committed:
missing short desc

roigcarlo committed:
Naming

roigcarlo committed:
Changed malloc for new and renamed variables as buffers

roigcarlo committed:
typo

roigcarlo committed:
Boost::weak_ptr and std::weak_ptr interfaces

roigcarlo committed:
Boost::weak_ptr tests and added a boost::shared_ptr test for const values

roigcarlo commented:
@pooyan-dadvand I have addressed your comments ( I hope all of them ). Just some small comments with the weak_ptr...

Also, about the rank value with single process runs. I am getting rank 0 systematically but I cannot find it in the doc yet.

pooyan-dadvand commented:
I have not seen the move constructor and assignement. Have you commit it? (Actually the default one would work for this case)

roigcarlo commented:
Ups your right, sorry. Lost in another branch, I'll upload them now

roigcarlo committed:
Moving = operator up and adding missing move constructor

roigcarlo commented:
@jcotela @RiccardoRossi @pooyan-dadvand 
There is another detail before merging. 
As you see there are tests for the MPI version. This has two problems:

1 - You cannot enable MPI in the kratos core at will because _CMakeList.txt_:
```C++
#finding Mpi

# VVVVVVVVVVVVVVVVV
set(MPI_NEEDED OFF)
# ^^^^^^^^^^^^^^^^^

if(${METIS_APPLICATION} MATCHES ON )
  set(MPI_NEEDED ON)
endif(${METIS_APPLICATION} MATCHES ON )
# etc...
```

That line prevents you from enabling it in the configure.

2 - Also, once the code compiles with the MPI enabled, as is part of Kratos core it need to link against mpi, which does not.

So, we were discussing with @pooyan-dadvand to either move the mpi test somewhere else, or make kratos link with MPI if MPI_NEEDED=ON.

Thoughts?

philbucher commented:
I think having a flag in the `configure.sh` for enabling MPI is much clearer. I also thought of suggesting this

RiccardoRossi reviewed (changes requested):
guys, i block this as i think that this point should be addressed (as we discussed on friday at lunch)
please take a look to my comments in the thread. in particular there is one that says the following:

using a "#ifdef MPI" makes impossible to use in the same executable MPI and OpenMP alone (without MPI).

i would love to keep that possibility... any way to prserve that?
to explain myself more precisely, the scenario i see is the following:

i am on a cluster, and i compile with mpi=0n. Now i want to run OpenMP alone. I will still need mpirun...while this is definitely not what i would want. An excellent example of this is to run the tests

roigcarlo commented:
@RiccardoRossi Why can't you run the program with MPI?. I was not there during the discussion you had about this.

Is this about the singleton MPI_INIT if compiled with mpich/pmi?

jcotela commented:
@roigcarlo The problem is as follows: assume I compiled my kratos with MPI support, but now I want to do an OpenMP run. If your object does special things when compiled with MPI, it will try to do them in my runs but, since I didn't call MPI_Init at any point, it will fail. Your only option then is to add MPI_Initialized calls to check for MPI status everywhere, which is very cumbersome.

pooyan-dadvand commented:
The only way that I see is to have the rank an static member of the Kernel: `Kernel::GetRank()` and having an MPIKernel which initialize it to the mpi one. (we are discussing this with @jcotela  for a while now)

jcotela commented:
@pooyan-dadvand and I were discussing this the other day and we agreed that we need to understand a bit better how to use separate communicators (for example: one mpi comm for the structure and another, non-overlapping one for the fluid) before deciding on the issue of determining the rank.

The main point is *what does the global pointer need the rank for*. Consider a scenario where ranks 0-9 solve the fluid and ranks 10-19 solve the structure. If we use global pointers within the structure, they can use the rank indices for MPI_COMM_WORLD (10-19) or the structure communicator rank indices (0-9).

Using 10-19 is good if we want to have *unique* indices for the pointers, but useless to exchange information using the structure communicator. Conversely, using 0-9 is good for communication within the structure but not unique (so a global pointer with a rank of 9 will refer to different processes depending on the communicator used).

As an aside, since I had to read a bit on these smaller communicators, I'm linking to a small [tutorial](http://mpitutorial.com/tutorials/introduction-to-groups-and-communicators/) I found useful.

roigcarlo committed:
Removed MPI dependent code

roigcarlo committed:
Move parallel test to a diferent file

roigcarlo committed:
Restored GetRank

roigcarlo committed:
Explicitly adding the rank in the constructor

roigcarlo committed:
Remove test_global_pointer_mpi from core

roigcarlo committed:
Cleaning up unused and potentially confusing includes

roigcarlo committed:
Decoupled bins_dynamic_objects_mpi from serial

Mpi version now uses a partition bins independent from the object bins
to handle the partition and a member object bins to handle the objects.
This prevents a very odd error that was causing templates to instantiate
to the point code instead of the object one

roigcarlo committed:
Adding test to the mpi application, and celaning up things

pooyan-dadvand commented on kratos/includes/global_pointer.h:
```diff
@@ -0,0 +1,217 @@
+class GlobalPointer {
+private:
+
+  TDataType GetPointer() { return mDataPointer; }
```
It mustn't be 

``` c++
TDataType* GetPointer() { return mDataPointer; }
```

pooyan-dadvand commented on kratos/includes/global_pointer.h:
```diff
@@ -0,0 +1,217 @@
+  TDataType * operator->() {
+    return mDataPointer;
+  }
+
```
missing the constant version of operator ->

RiccardoRossi replied:
@pooyan-dadvand this is done, i would say you could dismiss your review

pooyan-dadvand reviewed (changes requested):
Comments are given....

roigcarlo committed:
Added const arrow operator and deleted Getpointer()

pooyan-dadvand reviewed (dismissed):
For me it is ok, Just to comment that the common world is not stored inside which means that the caller should know in which world is. I'm remarking this because in our design with @jcotela we thought that it should store it. 

@jcotela do you have any comment? @RiccardoRossi is this ok for you?

RiccardoRossi commented on kratos/includes/global_pointer.h:
```diff
@@ -0,0 +1,222 @@
+  /// Pointer to the data
+  TDataType * mDataPointer;
+
+#ifdef KRATOS_USING_MPI
```
how will you know when you are within an appliction that needs mpi, that the core was compiled with mpi?

pooyan-dadvand commented:
@roigcarlo After discussing with @RiccardoRossi we arrived to conclusion that it would be better to remove `KRATOS_MPI` macros and leave global pointer with overhead of rank all the time.

Do you agree?

RiccardoRossi commented on applications/mpi_search_application/custom_python/add_custom_utilities_to_python.cpp:
```diff
@@ -73,7 +67,7 @@ namespace Kratos
+
             class_<MpiUtilitiesType, boost::noncopyable>
                     ("MpiUtilities", init<>())
                     .def("Repart",                  &MpiUtilitiesType::ParallelPartitioning)
```
minor: i don't think this is a very good name. Maybe "Repartition"? (just ignore the comment if you like the current one)

pooyan-dadvand replied:
Repartition is definitely a better name.

RiccardoRossi commented:
thx @pooyan-dadvand on my side that dissipates all of my concerns

philbucher commented:
hi guys do you have a plan of when to merge this?

pooyan-dadvand commented:
@roigcarlo I still see the `KRATOS_USING_MPI` macros. Do you agree in removing them?

roigcarlo commented:
Sure, we speak about removing this already but was never implemented

El lun., 1 ene. 2018 23:46, Pooyan Dadvand <notifications@github.com>
escribió:

> @roigcarlo <https://github.com/roigcarlo> I still see the KRATOS_USING_MPI
> macros. Do you agree in removing them?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/459#issuecomment-354682440>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB2Jr4MTr0IKS4pn1s2B8LUME0-T8BSxks5tGWBhgaJpZM4Nud4z>
> .
>

pooyan-dadvand commented:
so, would you please proceed with implementation?

roigcarlo committed:
Remove KRATOS_USING_MPI

roigcarlo commented:
@pooyan-dadvand Done

roigcarlo committed:
Remove KRATOS_USING_MPI

philbucher commented on kratos/includes/global_pointer.h:
```diff
@@ -63,10 +56,7 @@ class GlobalPointer {
    */
   GlobalPointer(boost::shared_ptr<TDataType> DataPointer, int Rank = 0)
```
`Kratos:: ...`?

roigcarlo replied:
Not sure what you mean, something wrong there?

philbucher replied:
nothing wrong, was just wondering why you don't use the new `Kratos::shared_ptr` => #1328

roigcarlo replied:
:+1:  But I will make the changes in a separate PR, since this is very old and if I merge the master with this it will crash

roigcarlo commented:
@RiccardoRossi or @pooyan-dadvand is this ok? can we merge this?

pooyan-dadvand commented on kratos/includes/global_pointer.h:
```diff
@@ -0,0 +1,192 @@
+   * Fills buffer with the GlobalPoiter data
+   * @param buffer Object data buffer
+   */
+  void Save(char * buffer) {
```
`save` is lower case in all Kratos.

pooyan-dadvand replied:
Please apply this comment

pooyan-dadvand commented on kratos/includes/global_pointer.h:
```diff
@@ -0,0 +1,192 @@
+   * Restores the GlobalPoiter with the data from the buffer
+   * @param buffer Object data buffer
+   */
+  void Load(char * buffer) {
```
`load` is lower case in all Kratos.

pooyan-dadvand replied:
here the same

pooyan-dadvand commented on kratos/tests/global_pointers/test_global_pointer.cpp:
```diff
@@ -0,0 +1,197 @@
+// Boost::shared_ptr
+KRATOS_TEST_CASE_IN_SUITE(GlobalPointerCreateBoostSharedPtr, KratosCoreFastSuit)
+{
+  typedef boost::shared_ptr<TestClass> BoostPtrType;
```
We should change this to Kratos:: (maybe afterward)

roigcarlo replied:
Yep. As I said to @philbucher I will do this once its merged

pooyan-dadvand reviewed (changes requested):
More comments...

roigcarlo committed:
Fixing function names

roigcarlo commented:
@pooyan-dadvand According to @RiccardoRossi this has more priority now. Honestly I did not remember the status of this but seems to be finished, but we cannot merge it because this is prior to pybind etc... I will adapt it so we can merge if you agree

loumalouomega commented:
@philbucher and I will be very happy of this. With this the contact and mapping implementation will be easier on MPI

philbucher commented:
I recently started implementing a more general mapper, but without using this.
I will reevaluate if I can use it once it is merged

RiccardoRossi reviewed (commented):
just to say that looking at this today, with the rank being passed always and being equal to zero in openmp case i am quite ok! It does not make sense to approve it now, since it will simply not compile, however i am positive on this

roigcarlo committed:
Merge branch 'master' into feature-global-pointers

roigcarlo committed:
Porting to pybind

roigcarlo committed:
Removing references to boost::pointers

roigcarlo committed:
Removing duplicate constructors

roigcarlo committed:
Indentation

roigcarlo committed:
Boost:: -> Kratos::

roigcarlo committed:
Merge branch 'master' into feature-global-pointers

roigcarlo commented:
@pooyan-dadvand This should be usable now.

roigcarlo committed:
Fixing string cmp warning

RiccardoRossi reviewed (dismissed):
this looks fine to me, apart for not having any test for mpi search (i think at least)

I am approving...however if you could find the time to add a small mpi search test it would be perfect

pooyan-dadvand reviewed (commented):
There are still some minor naming inconsistencies to be finished.

roigcarlo committed:
Changing Repartition -

roigcarlo commented:
@pooyan-dadvand I've changed the Repartiton one, but I think you were reviewing an old version of the global_pointer.h file....

pooyan-dadvand reviewed (commented):
And load and save please

pooyan-dadvand commented:
Now I see, they were changed!
